### PR TITLE
Convert blacklisted IPv4 addresses to compatible IPv6 addresses.

### PR DESCRIPTION
**Pull request by clokep**
This automatically converts IPv4 addresses to some obscure IPv6 address ranges:

* IPv6 with embedded IPv4
* IPv4-compatible IPv6
* 6to4 addresses

This expands on some of the conversation in https://github.com/matrix-org/synapse/pull/8870#discussion_r535485437.

~~An alternative approach to this would be to automatically do these conversions for any IPv4 addresses that get blocked at runtime, not just for the default.~~ This approach is now taken.

### CODE REVIEW

**Comment by clokep** [comment]
> An alternative approach to this would be to automatically do these conversions for any IPv4 addresses that get blocked at runtime, not just for the default.

From some chatter I think we're going to switch to this instead.

File: synapse/config/repository.py [hunk 1]
```diff
@@ -187,15 +185,13 @@ def read_config(self, config, **kwargs):
...
 
-            self.url_preview_ip_range_blacklist = IPSet(
-                config["url_preview_ip_range_blacklist"]
-            )
-
             # we always blacklist '0.0.0.0' and '::', which are supposed to be
             # unroutable addresses.
-            self.url_preview_ip_range_blacklist.update(["0.0.0.0", "::"])
+            self.url_preview_ip_range_blacklist = generate_ip_set(
+                config["url_preview_ip_range_blacklist"], ["0.0.0.0", "::"]
```
Comment (clokep) [inline synapse/config/repository.py]: I wonder if these default IPs should be CIDRs?

File: synapse/config/server.py [hunk 2]
```diff
@@ -294,32 +355,28 @@ def read_config(self, config, **kwargs):
...
         # Attempt to create an IPSet from the given ranges
-        try:
-            self.ip_range_blacklist = IPSet(ip_range_blacklist)
-        except Exception as e:
-            raise ConfigError("Invalid range(s) provided in ip_range_blacklist.") from e
+
         # Always blacklist 0.0.0.0, ::
-        self.ip_range_blacklist.update(["0.0.0.0", "::"])
+        self.ip_range_blacklist = generate_ip_set(
+            ip_range_blacklist, ["0.0.0.0", "::"], config_path=("ip_range_blacklist",)
```
Comment (clokep) [inline synapse/config/server.py:361]: I wonder if these default IPs should be CIDRs?

File: synapse/config/server.py [hunk 3]
```diff
@@ -40,6 +41,66 @@
...
+        try:
+            network = IPNetwork(ip)
+        except AddrFormatError as e:
+            raise ConfigError(
+                "Invalid IP range provided: %s." % (ip,), config_path
+            ) from e
+        result.add(network)
+
+        # It is possible that these already exist in the set, but that's OK.
+        if ":" not in str(network):
```
Comment (erikjohnston) [inline synapse/config/server.py:95]: *How* is there not a `network.is_ipv4()`???

File: synapse/config/server.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/config/server.py:361]: I think so? Probably?

**Review by erikjohnston — APPROVED** [review]

File: synapse/config/server.py (hunk 3 above)
Comment (clokep) [reply synapse/config/server.py:95]: I think because every IPv4 network is technically an IPv6 network too since all of IPv4 is included in IPv6? 🤷

Comment (clokep) [reply synapse/config/server.py:95]: So really all of this works around a bug in netaddr that it doesn't understand the different forms of IPv6 addresses really encompass the same networks as the IPv4 ones. 🤷

Comment (erikjohnston) [reply synapse/config/server.py:95]: :shrug:

File: synapse/config/server.py (hunk 2 above)
Comment (clokep) [reply synapse/config/server.py]: Updated 0.0.0.0 -> 0.0.0.0/8 which matches the default blacklist.

**Comment by clokep** [comment]
Based on https://en.wikipedia.org/wiki/Reserved_IP_addresses I also added a few more IP ranges for testing networks and IPv6 multicast (which doesn't make sense).

File: synapse/config/server.py [hunk 4]
```diff
@@ -66,6 +129,10 @@
     "fe80::/10",
     # Unique local addresses.
     "fc00::/7",
+    # Testing networks.
+    "2001:db8::/32",
+    # Multicast.
+    "ff00::/8",
```
Comment (dkasak) [inline synapse/config/server.py:135]: Let's also add `fec0::/10:` Site Local Addresses (deprecated - RFC 3879), previously mentioned [here](https://github.com/matrix-org/synapse/pull/8870#discussion_r536253356).

```suggestion
    "ff00::/8",
    # Site-local addresses
    "fec0::/10",
```

File: docs/sample_config.yaml [hunk 5]
```diff
@@ -177,6 +178,8 @@ pid_file: DATADIR/homeserver.pid
 #  - '::1/128'
 #  - 'fe80::/10'
 #  - 'fc00::/7'
+#  - '2001:db8::/32'
+#  - 'ff00::/8'
```
Comment (dkasak) [inline docs/sample_config.yaml:182]:
```suggestion
#  - 'ff00::/8'
#  - 'fec0::/10'
```

File: docs/sample_config.yaml [hunk 6]
```diff
@@ -1002,6 +1006,8 @@ media_store_path: "DATADIR/media_store"
...
```
Comment (dkasak) [inline docs/sample_config.yaml:1010]:
```suggestion
#  - 'ff00::/8'
#  - 'fec0::/10'
```

**Review by dkasak — CHANGES_REQUESTED** [review]

**Comment by clokep** [comment]
@dkasak / @erikjohnston Are we happy with the current bit of this PR? I think it is in good shape now.

**Review by erikjohnston — APPROVED** [review]
Code looks fine, I think the ip ranges makes sense, but let's wait for @dkasak to confirm

File: synapse/config/repository.py [hunk 7]
```diff
@@ -187,16 +185,17 @@ def read_config(self, config, **kwargs):
...
+                config["url_preview_ip_range_blacklist"],
+                ["0.0.0.0", "::"],
+                config_path=("url_preview_ip_range_blacklist",),
+            )
 
-            self.url_preview_ip_range_whitelist = IPSet(
-                config.get("url_preview_ip_range_whitelist", ())
+            self.url_preview_ip_range_whitelist = generate_ip_set(
+                config.get("url_preview_ip_range_whitelist", ()),
+                config_path=("url_preview_ip_range_blacklist",),
```
Comment (dkasak) [inline synapse/config/repository.py:198]: Shouldn't this be `"url_preview_ip_range_whitelist"`?

```suggestion
                config_path=("url_preview_ip_range_whitelist",),
```

File: synapse/config/server.py [hunk 8]
```diff
@@ -40,6 +41,66 @@
...
+    """
+    Generate an IPSet from a list of IP addresses or CIDRs.
+
+    Additionally, for each IPv4 network in the list of IP addresses, also
+    includes the corresponding IPv6 networks.
+
+    This includes:
+
+    * IPv4-Compatible IPv6 Address (see RFC 4291, section 2.5.5.1)
+    * IPv4-Mapped IPv6 Address (see RFC 4291, section 2.5.5.1)
```
Comment (dkasak) [inline synapse/config/server.py:73]:
```suggestion
    * IPv4-Mapped IPv6 Address (see RFC 4291, section 2.5.5.2)
```

**Review by dkasak — CHANGES_REQUESTED** [review]
Looks good to me too, apart from two minor things.

(do we Approve or Request changes in such scenarios?)

**Comment by clokep** [comment]
> Looks good to me too, apart from two minor things.

Good catches!

> (do we Approve or Request changes in such scenarios?)

My rule of thumb is request changes if you want to look at them again before they get merged; approve if you're OK with the author making the changes and merging.

**Review by dkasak — APPROVED** [review]

**Pull request merged**